source/common/formatter/substitution_formatter.cc

```diff
@@ -1987,6 +1994,62 @@ UpstreamHostMetadataFormatter::UpstreamHostMetadataFormatter(const std::string&
                           return host->metadata().get();
                         }) {}
 
+ProxyProtocolTlvsFormatter::ProxyProtocolTlvsFormatter(const std::string& tlv_type_str) {
+  // Specified tlv_type must be parsable as an int.
+  ASSERT(!tlv_type_str.empty());
+  if (!absl::SimpleAtoi(tlv_type_str, &tlv_type_)) {
+    throw EnvoyException(fmt::format(
+        "Invalid parameter provided for PROXY_PROTOCOL_TLVS header: {}. Not parsable as int.",
+        tlv_type_str));
+  }
+
+  // Check if a valid TLV type was passed in.
+  if (tlv_type_ >= 256 || tlv_type_ <= 0) {
+    throw EnvoyException(fmt::format("Invalid parameter provided for PROXY_PROTOCOL_TLVS header: "
+                                     "{}. Must be a positive integer less than 256.",
+                                     tlv_type_str));
+  }
+}
+
+absl::optional<std::string> ProxyProtocolTlvsFormatter::format(
+    const Http::RequestHeaderMap&, const Http::ResponseHeaderMap&, const Http::ResponseTrailerMap&,
+    const StreamInfo::StreamInfo& stream_info, absl::string_view) const {
+  const auto& typed_state =
+      stream_info.filterState().getDataReadOnly<Network::ProxyProtocolFilterState>(
+          Network::ProxyProtocolFilterState::key());
+
+  // ProxyProtocolFilterState is not stored in the filter state.
+  if (typed_state == nullptr) {
+    ENVOY_LOG_MISC(debug, "Invalid header: PROXY_PROTOCOL_TLVS. The ProxyProtocolFilterState is "
+                          "not stored in the filter state.");
+    return absl::nullopt;
+  }
+
+  std::ostringstream oss;
+  int match_count = 0;
+  for (auto& tlv : typed_state->value().tlv_vector_) {
+    if (tlv.type == tlv_type_) {
+      if (match_count > 0)
+        oss << ", ";
+      oss << Base64::encode(reinterpret_cast<const char*>(tlv.value.data()), tlv.value.size());
```

Comment:
[note] This reinterpret cast is the change to the original patch made for SRE from the work done in [this commit](https://github.com/10gen/envoy-serverless/commit/b63805eef506087fdbe9533dc084530ce16d22eb). `Base64::encode` requires the `const char*` type, not `const unsigned char*`. 

When Siyuan's changes merged upstream, a previously `std::string` value type was modified to be `std::vector<unsigned char>`, so we're now required to make this cast ([upstream code link](https://github.com/envoyproxy/envoy/blob/c27c885a08318506915b0e90f0bb214e76505e3d/envoy/network/proxy_protocol.h#L14), [current patch used by SRE](https://github.com/xgen-cloud/rpm-spec/blob/db1c3909c62d9f1edb39beee0fb96d1a065dcdf9/build/envoy/SOURCES/serverless-pass-through-tlvs.patch#L19))